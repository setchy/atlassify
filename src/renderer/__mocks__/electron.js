// /**
//  * @typedef {import('../../preload/types.ts').AtlassifyAPI} AtlassifyAPI
//  */

// @ts-ignore
window.Notification = function (title) {
  this.title = title;

  return {
    onclick: jest.fn(),
  };
};

// @ts-ignore
window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

// @ts-ignore
window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: jest.fn(),
};

window.alert = jest.fn();

// /** @type {AtlassifyAPI} */
// const mockAtlassifyAPI = {
// 	openExternalLink: jest.fn(),
// 	decryptValue: () => Promise.resolve("decrypted"),
// 	encryptValue: () => Promise.resolve("encrypted"),
// 	platform: {
// 		isLinux: () => false,
// 		isMacOS: () => true,
// 		isWindows: () => false,
// 	},
// 	app: {
// 		hide: jest.fn(),
// 		show: jest.fn(),
// 		quit: jest.fn(),
// 		version: () => Promise.resolve("0.0.1"),
// 	},
// 	zoom: {
// 		getLevel: jest.fn(),
// 		setLevel: jest.fn(),
// 	},
// 	tray: {
// 		updateIcon: jest.fn(),
// 		updateTitle: jest.fn(),
// 		useAlternateIdleIcon: jest.fn(),
// 	},
// 	twemojiDirectory: jest.fn(),
// };

// window.atlassify = mockAtlassifyAPI;

// window.atlassify.decryptValue = Promise.resolve("decrypted");
// window.atlassify.twemojiDirectory = Promise.resolve("/tmp/emoji");

window.atlassify.app = {
  version: Promise.resolve('0.0.1'),
};
