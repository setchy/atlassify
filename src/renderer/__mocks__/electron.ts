import { namespacedEvent } from '../../shared/events';

// Mock `window.Notification`
vi.stubGlobal('Notification', function (title: string) {
  this.title = title;
  return {
    onclick: vi.fn(),
  };
});

// Mock `window.Audio`
vi.stubGlobal(
  'Audio',
  class Audio {
    path: string;
    constructor(path: string) {
      this.path = path;
    }
    play() {}
  },
);

// Mock `window.localStorage`
vi.stubGlobal('localStorage', {
  store: {} as Record<string, string>,
  getItem: vi.fn((key: string) => localStorage.store[key] || null),
  setItem: vi.fn((key: string, item: string) => {
    localStorage.store[key] = item;
  }),
  removeItem: vi.fn(),
});

// Mock `window.alert`
vi.stubGlobal('alert', vi.fn());

export const ipcRenderer = {
  send: vi.fn(),
  on: vi.fn(),
  sendSync: vi.fn(),
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  invoke: vi.fn((channel: string, ..._args: any[]) => {
    switch (channel) {
      case 'get-platform':
        return Promise.resolve('darwin');
      case namespacedEvent('version'):
        return Promise.resolve('0.0.1');
      case namespacedEvent('safe-storage-encrypt'):
        return Promise.resolve('encrypted');
      case namespacedEvent('safe-storage-decrypt'):
        return Promise.resolve('decrypted');
      default:
        return Promise.reject(new Error(`Unknown channel: ${channel}`));
    }
  }),
};

export const shell = {
  openExternal: vi.fn(),
};

export const webFrame = {
  setZoomLevel: vi.fn(),
  getZoomLevel: vi.fn(),
};
