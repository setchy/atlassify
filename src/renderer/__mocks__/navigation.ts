const mockNavigate = vi.fn();

vi.mock('react-router-dom', async () => ({
  ...(await vi.importActual('react-router-dom')),
  useNavigate: () => mockNavigate,
}));

export { mockNavigate };
