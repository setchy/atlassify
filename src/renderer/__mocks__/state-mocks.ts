import {
	type Account,
	type AppearanceSettingsState,
	type AtlassifyState,
	type AuthState,
	type EncryptedToken,
	type FilterSettingsState,
	type Link,
	type NotificationSettingsState,
	OpenPreference,
	type SettingsState,
	type SystemSettingsState,
	Theme,
	type Username,
} from "../types";

export const mockAtlassianCloudAccount: Account = {
	id: "123456789",
	username: "user@atlassify.io" as Username,
	token: "token-123-456" as EncryptedToken,
	name: "Atlassify",
	avatar: "https://avatar.atlassify.io" as Link,
};

export const mockAuth: AuthState = {
	accounts: [mockAtlassianCloudAccount],
};

const mockAppearanceSettings: AppearanceSettingsState = {
	theme: Theme.LIGHT,
	zoomPercentage: 100,
	language: "en",
};

const mockNotificationSettings: NotificationSettingsState = {
	markAsReadOnOpen: true,
	delayNotificationState: false,
	fetchOnlyUnreadNotifications: true,
	groupNotificationsByProduct: false,
	groupNotificationsByProductAlphabetically: false,
};

const mockSystemSettings: SystemSettingsState = {
	openLinks: OpenPreference.FOREGROUND,
	keyboardShortcutEnabled: true,
	showNotificationsCountInTray: true,
	showSystemNotifications: true,
	playSoundNewNotifications: true,
	useAlternateIdleIcon: false,
	openAtStartup: true,
};

const mockFilters: FilterSettingsState = {
	filterTimeSensitive: [],
	filterCategories: [],
	filterReadStates: [],
	filterProducts: [],
	filterActors: [],
};

export const mockSettings: SettingsState = {
	...mockAppearanceSettings,
	...mockNotificationSettings,
	...mockSystemSettings,
	...mockFilters,
};

export const mockState: AtlassifyState = {
	auth: mockAuth,
	settings: mockSettings,
};
