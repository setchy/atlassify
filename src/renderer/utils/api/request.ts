import axios, { type Method, type AxiosPromise } from 'axios';
import type { Account, Token, Username } from '../../types';
import { decryptValue } from '../comms';
import { Constants } from '../constants';
import type { TypedDocumentString } from './gql/graphql';
import type { GraphQLRequest } from './types';

export async function performPostRequest(
  account: Account,
  data: GraphQLRequest,
): AxiosPromise {
  // TODO consider storing the decrypted token in memory
  const decryptedToken = await decryptValue(account.token);
  const auth = btoa(`${account.username}:${decryptedToken}`);

  return performApiRequest(auth, 'POST', data);
}

export function performHeadRequest(
  username: Username,
  token: Token,
): AxiosPromise | null {
  const auth = btoa(`${username}:${token}`);

  return performApiRequest(auth, 'HEAD');
}

function performApiRequest(
  auth: string,
  method: Method,
  data = {},
): AxiosPromise | null {
  const url = Constants.ATLASSIAN_URLS.API;

  axios.defaults.headers.common.Accept = '*/*';
  axios.defaults.headers.common.Authorization = `Basic ${auth}`;
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  axios.defaults.headers.common['Content-Type'] = 'application/json';

  return axios({ method, url, data });
}

export async function execute<TResult, TVariables>(
  account: Account,
  query: TypedDocumentString<TResult, TVariables>,
  ...[variables]: TVariables extends Record<string, never> ? [] : [TVariables]
) {
  const decryptedToken = await decryptValue(account.token);
  const auth = btoa(`${account.username}:${decryptedToken}`);

  const response = await fetch(Constants.ATLASSIAN_URLS.API, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: '*/*',
      Authorization: `Basic ${auth}`,
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  if (!response.ok) {
    throw new Error('Network response was not ok');
  }

  return response.json() as TResult;
}
