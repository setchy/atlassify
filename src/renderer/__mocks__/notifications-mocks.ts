import type {
  AccountNotifications,
  AtlassifyNotification,
  Link,
} from '../types';
import { PRODUCTS } from '../utils/products';
import { mockAtlassianCloudAccount } from './state-mocks';

export const mockAtlassifyNotifications: AtlassifyNotification[] = [
  {
    id: '138661096',
    order: 0,
    notificationGroup: {
      id: '138661096',
      size: 1,
      additionalActors: [],
    },
    message: '#103: chore(deps): update dependency eslint',
    updated_at: '2020-09-13T21:12:16.662Z',
    entity: {
      title: '#103: chore(deps): update dependency eslint',
      iconUrl:
        'https://bbc-object-storage--frontbucket.us-east-1.prod.public.atl-paas.net/80907f89f58b/img/favicon.ico' as Link,
      url: 'https://bitbucket.org/myorg/notifications-test/pull-requests/103?link_source=platform' as Link,
    },
    path: null,
    product: PRODUCTS.bitbucket,
    actor: {
      displayName: 'atlassify-app',
      avatarURL:
        'https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/123' as Link,
    },
    category: 'direct',
    readState: 'unread',
    account: mockAtlassianCloudAccount,
    url: 'https://bitbucket.org/myorg/notifications-test/pull-requests/103?link_source=platform' as Link,
    type: 'unknown',
  },
  {
    id: '148827438',
    order: 1,
    notificationGroup: {
      id: '138661096',
      size: 2,
      additionalActors: [
        {
          displayName: 'atlassify-app',
          avatarURL:
            'https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/123' as Link,
        },
      ],
    },
    message: 'atlassify-app edited your page',
    updated_at: '2020-09-13T21:12:16.662Z',
    entity: {
      title: 'Atlassify Home',
      iconUrl:
        'https://home-static.us-east-1.prod.public.atl-paas.net/confluence-page-icon.svg' as Link,
      url: 'https://some-tenant.atlassian.net/wiki/spaces/Atlassify/pages/3752329340/Atlassify+Home' as Link,
    },
    path: {
      title: 'Atlassify Space',
      iconUrl: null,
      url: 'https://some-tenant.atlassian.net/wiki/spaces/Atlassify' as Link,
    },
    product: PRODUCTS.confluence,
    actor: {
      displayName: 'atlassify-app',
      avatarURL:
        'https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/123' as Link,
    },
    category: 'direct',
    readState: 'unread',
    account: mockAtlassianCloudAccount,
    url: 'https://some-tenant.atlassian.net/wiki/spaces/Atlassify/pages/3752329340/Atlassify+Home' as Link,
    type: 'Create',
  },
];

export const mockSingleAtlassifyNotification: AtlassifyNotification =
  mockAtlassifyNotifications[0];

export const mockAccountNotifications: AccountNotifications[] = [
  {
    account: mockAtlassianCloudAccount,
    notifications: mockAtlassifyNotifications,
    hasMoreNotifications: false,
    error: null,
  },
];

export const mockAccountNotificationsWithMorePages: AccountNotifications[] = [
  {
    account: mockAtlassianCloudAccount,
    notifications: mockAtlassifyNotifications,
    hasMoreNotifications: true,
    error: null,
  },
];

export const mockSingleAccountNotifications: AccountNotifications[] = [
  {
    account: mockAtlassianCloudAccount,
    notifications: [mockAtlassifyNotifications[0]],
    hasMoreNotifications: false,
    error: null,
  },
];
