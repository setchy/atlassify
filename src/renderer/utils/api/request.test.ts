import axios from 'axios';

import { mockAtlassianCloudAccount } from '../../__mocks__/state-mocks';
import type { Link } from '../../types';
import type { MeQuery, TypedDocumentString } from './graphql/generated/graphql';
import { performHeadRequest, performPostRequest } from './request';

jest.mock('axios');

const url = 'https://team.atlassian.net/gateway/api/graphql' as Link;

describe('renderer/utils/api/request.ts', () => {
  beforeEach(() => {
    (axios as jest.MockedFunction<typeof axios>).mockResolvedValue({
      data: {
        data: {},
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should execute graphql request with the correct parameters', async () => {
    const data = {
      query: 'foo',
      variables: {
        key: 'value',
      },
    };
    const method = 'POST';

    await performPostRequest(
      mockAtlassianCloudAccount,
      'foo' as unknown as TypedDocumentString<MeQuery, unknown>,
      {
        key: 'value',
      },
    );

    expect(axios).toHaveBeenCalledWith({
      method,
      url,
      data,
    });

    expect(axios.defaults.headers.common).toMatchSnapshot();
  });

  it('should make a head request with the correct parameters', async () => {
    const data = {};
    const method = 'HEAD';

    await performHeadRequest(
      mockAtlassianCloudAccount.username,
      mockAtlassianCloudAccount.token,
    );

    expect(axios).toHaveBeenCalledWith({
      method,
      url,
      data,
    });

    expect(axios.defaults.headers.common).toMatchSnapshot();
  });
});
