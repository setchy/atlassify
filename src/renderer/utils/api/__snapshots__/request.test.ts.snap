// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderer/utils/api/request.ts > should execute graphql request with the correct parameters 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/request.ts > should make a head request with the correct parameters 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86dG9rZW4tMTIzLTQ1Ng==",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;
